src/compiler/scala/tools/nsc/typechecker/Checkable.scala

```diff
@@ -179,20 +162,48 @@ trait Checkable {
       |[P4] $P4%-6s None of the above  // !(P1 || P2 || P3)
     """.stripMargin.trim
 
-    val result = (
+    val result: Checkability =
       if (X.isErroneous || P.isErroneous) CheckabilityError
       else if (P1) StaticallyTrue
       else if (P2) StaticallyFalse
       else if (P3) RuntimeCheckable
-      else if (uncheckableType == NoType) {
-        // Avoid warning (except ourselves) if we can't pinpoint the uncheckable type
-        debuglog("Checkability checker says 'Uncheckable', but uncheckable type cannot be found:\n" + summaryString)
-        CheckabilityError
+      else if (uncheckableType != NoType) Uncheckable
+      else // Avoid warning (except ourselves) if we can't pinpoint the uncheckable type
+        CheckabilityError.tap(_ => debuglog(s"Checkability checker says 'Uncheckable', but uncheckable type cannot be found:\n$summaryString"))
+        //CheckabilityError.tap(_ => println(s"Checkability checker says 'Uncheckable', but uncheckable type cannot be found:\n$summaryString"))
+    // collect type args which are candidates for warning because uncheckable
+    private def typeArgsInTopLevelType(tp: Type): Set[Type] = {
+      def isUnwarnableTypeArg(arg: Type) = {
+        //import util._
+        def isUnwarnableTypeArgSymbol(sym: Symbol) = {
+          @`inline` def exemption = sym.owner == tp.typeSymbol || inPattern && (sym.owner.name == nme.unapply || sym.owner.name == nme.unapplySeq)
+          sym.isTypeParameter && exemption ||     // dummy
```

Comment:
OK, so `dummy` refers to https://github.com/scala/scala/blob/37cf33ac78aee3ce457b638de62106c4c3be4dd4/src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala#L394, and `exemption` is trying to capture this more precisely?

Maybe we can find a better secret handshake between the two methods?

In this example

```
class Renderer[T]
class Test[A](a: Any) {
  def f2[B] = a match { case _: Renderer[B] => }
  def g = a match { case _: Renderer[A] => }
}
```

I don't get why `sym.isTypeParameter` is true for the `A` in `case _: Renderer[A]`, but it's false for the `B` in `case  _: Renderer[B]`. I guess it's because I still don't get type parameter skolems.

Reply:
I looked a bit at what might differentiate syms here vs passing flags down the stack for more context, or add attachment somewhere. I pushed an update that is still WIP. But I don't have a complete picture yet.